mlir/lib/Ion/Transforms/oqd_database_managers.hpp

```diff
@@ -0,0 +1,93 @@
+// Copyright 2024 Xanadu Quantum Technologies Inc.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//     http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <cassert>
+#include <regex>
+#include <type_traits>
+
+#include <toml++/toml.hpp>
+
+static const std::string catalyst_root_path = std::regex_replace(
+    __FILE__, std::regex("mlir/lib/Ion/Transforms/./oqd_database_managers.hpp"), "");
+static const std::string oqd_device_parameters_toml_file_path =
+    catalyst_root_path + "frontend/catalyst/third_party/oqd/src/oqd_device_parameters.toml";
+static const std::string oqd_qubit_parameters_toml_file_path =
+    catalyst_root_path + "frontend/catalyst/third_party/oqd/src/oqd_qubit_parameters.toml";
+
+static const std::string oqd_gate_decomposition_parameters_toml_file_path =
+    catalyst_root_path +
+    "frontend/catalyst/third_party/oqd/src/oqd_gate_decomposition_parameters.toml";
+
+toml::parse_result load_toml_file(const std::string &path) { return toml::parse_file(path); }
+
+template <typename T> std::vector<T> TomlArray2StdVector(toml::array arr)
+{
+    std::vector<T> vec;
+    for (size_t i = 0; i < arr.size(); i++) {
+        if constexpr (std::is_same_v<T, int64_t>) {
+            vec.push_back(arr[i].as_integer()->get());
+        }
+        else if constexpr (std::is_same_v<T, double>) {
+            vec.push_back(arr[i].as_floating_point()->get());
+        }
+    }
```

Comment:
We should prefer range-based for loops. Also, I think we should pull the type-checking `if` statements out of the loop since we don't need to check the type in each iteration:

```suggestion
    if constexpr (std::is_same_v<T, int64_t>) {
        for (const auto &elem : arr) {
            vec.push_back(elem.as_integer()->get());
        }
    } 
    else if constexpr (std::is_same_v<T, double>) {
        for (const auto &elem : arr) {
            vec.push_back(elem.as_floating_point()->get());
        }
    }
```

Reply:
[Optional] Should we raise some sort of error if the type is not an `int64_t` or a `double`?

Reply:
`constexpr` only happens once at compile time (compile time of actually compiling the pass, i.e. once during `make dialects`, not compile time of qjit)

However I will make this change for readability!

Reply:
True, in that case feel free to leave the `if constexpr`s in the loop iteration, that way we don't have to write two for loops. I think we should still change this to a range-based for loop, however, e.g.

```suggestion
    for (const auto &elem : arr) {
        if constexpr (std::is_same_v<T, int64_t>) {
            vec.push_back(elem.as_integer()->get());
        }
        else if constexpr (std::is_same_v<T, double>) {
            vec.push_back(elem.as_floating_point()->get());
        }
    }
```

Reply:
done

Reply:
FYI on constexpr https://stackoverflow.com/questions/43434491/difference-between-if-constexpr-vs-if